### PR TITLE
MM-18062 - add support for Office365Settings Directory (tenant) Id

### PR DESCRIPTION
**Pull request by catalintomai**
#### Summary
Add support for setting Office365Settings Directory (tenant) Id in config.json.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-13514
https://mattermost.atlassian.net/browse/MM-18062
https://mattermost.atlassian.net/browse/MM-22376

#### Related Pull Requests
- Has webapp changes : https://github.com/mattermost/mattermost-webapp/pull/4717

### CODE REVIEW

File: model/config.go [hunk 1]
```diff
@@ -2682,6 +2739,10 @@ func (o *Config) IsValid() *AppError {
 		return err
 	}
 
+	if *o.Office365Settings.Enable && len(*o.Office365Settings.DirectoryId) == 0 {
+		return NewAppError("Config.IsValid", "model.oauth.is_valid.directory_id.app_error", nil, "", http.StatusBadRequest)
+	}
+
```
[the remark this answers is not in the document]
Comment (catalintomai) [reply model/config.go]: I added code to distinguish between a clean install (new config) and an upgrade (updating existing config).

**Comment by catalintomai** [comment]
> Had tested the PR and everything works fine. Just a small observation.
> 
> In the System Console > OAuth2.0 Page, the default value shown for the new field `Directory (tenant) ID` is `{directoryId}`. Can we instead leave this empty just like other fields.

Unfortunately, no, because with this change we also added validation on Save in the UI that the DirectoryId field is being populated. That means that we updated config's IsValid() method with new functionality that validates that this field (DirectoryId) is populated. Since the config's IsValid()method is called by both UI at save time and at server start time, if we would leave it empty, the config file validation would fail at server start time. The other fields do not have this validation currently and that is why they can be empty (meaning you can have no value and still save- see applicationID for example). This should go away once we redo the validation of fields in the SystemConsole UI.

**Comment by srkgupta** [comment]
@catalintomai Currently there is no such validation for Application ID and Application Secret. So why is there a special validation only for DirectoryId? Shouldn't the behaviour be consistent and let this field be empty. The Office365 login would not work even if one of the Application ID or Application Secret is empty. So why do we have to prevent start of the server if the field DirectoryId is empty?

**Comment by srkgupta** [comment]
@thefactremains can you please review the current behavior and confirm if this is fine. 

* On a new instance, the Office365 setting looks like this:
If you notice the fields Application ID and Application Secret are empty, however the field Directory (tenant) ID has a default value `{directoryId}` which needs to be cleared and replaced. 

Screenshot 2020-02-06 at 10 15 02 PM

* Currently there is a validation in place for the new field `Directory (tenant) ID` in config.json. If this field is empty, the server refuses to start. Should we have such a validation in place first of all? If this field is empty, then only Login via Office365 functionality will not work. Why do we have to prevent the rest of the app from being used. 

* Currently there are no validations in place if Application ID or Application Secret is left empty in the form. However we do have validation for the new field `Directory (tenant) ID`. Ideally we should have validations for all the fields, since the Login functionality will not work even if one of these field is empty.

**Comment by catalintomai** [comment]
> @catalintomai Currently there is no such validation for Application ID and Application Secret. So why is there a special validation only for DirectoryId? Shouldn't the behaviour be consistent and let this field be empty. The Office365 login would not work even if one of the Application ID or Application Secret is empty. So why do we have to prevent start of the server if the field DirectoryId is empty?

The fact that there is no validation for the other fields is a bug, but outside of the scope of the current change(both in UI and in config). When the server starts, config gets loaded and verifies that each Settings module of the config is valid. For O365 Settings, if enabled, it should check that the values are being set for the fields (right here I added this validation for DirectoryId, but we should add it for the other fields as well - this is the first step). Short of all of them being populated, having O365Settings "on" (or being able to save from the UX) doesn't really have a meaning (of course, the values can be present but be wrong but that is a different issue).

**Comment by thefactremains** [comment]
@catalintomai Since we are not currently planning to build validation with the same consistency across the System Console in the same release as this feature, I do not feel good about making this change to this single field. Especially considering this is not a high priority fix.

**Comment by catalintomai** [comment]
> Since we are not currently planning to build validation with the same consistency across the System Console in the same release as this feature, I do not feel good about making this change to this single field. Especially considering this is not a high priority fix.
> 
> To clarify, I am saying we cannot be inconsistent. If we do this formatting for one field, we should do it for all. If we cannot do it for all in the same release, we should not do it for one. It will be confusing and will make the product appear unprofessional.

@thefactremains , @srkgupta : Thet ype of validation present here already exists in the SysConsole, for example LDAP (and probably other places). If one goes and enables LDAP ("Enable sign-in with AD/LDAP" set to true) in the config file LDAP Settings section and then does not set one of the required fields (e.g. AD/LDAP Server) and then tries to start the MM server, the server does not start (also cannot save the config from the UX).

I agree on the inconsistence aspect, however, OTOH, it is a good practice to limit fixes/feature work to the scope of the ticket (not over-reach), one of the main reasons being the possibility of introducing regressions in areas supposedly not covered by the ticket + adding confusion to test (missing to test things because they are more or less surreptitiously introduced as part of some other work).
That being said, if we want to fix this I believe the right course of action is to open a ticket to fix validation for the other fields for O365 settings and make this ticket dependent on that one (or not).

**Comment by srkgupta** [comment]
@catalintomai Please note that incase of LDAP, we especially have a field which indicates if LDAP feature is enabled or not. Something like this:
Screenshot 2020-02-07 at 9 27 45 PM

However we dont have such field for Office365 setting on the System console UI. Hence if the configuration is invalid, it should ideally mean that office365 should be disabled and vice versa.

**Comment by catalintomai** [comment]
> @catalintomai Please note that incase of LDAP, we especially have a field which indicates if LDAP feature is enabled or not. Something like this:
> Screenshot 2020-02-07 at 9 27 45 PM
> 
> However we dont have such field for Office365 setting on the System console UI. Hence if the configuration is invalid, it should ideally mean that office365 should be disabled and vice versa.

@srkgupta - However, the O365Settings section in the config has "Enable": true field, like LDAP, even if not surfaced using a button in the UX. Haven't tried, however, my understanding is that the moment one selects the O365Settings in the dropdown at the top and clicks Save, it is enabled in the config O365Settings.

**Comment by srkgupta** [comment]
@catalintomai as long as we are handling the validations (UI & Server) as part of the same release as this feature, I am fine with it. I have raised this issue separately here:
https://mattermost.atlassian.net/browse/MM-22376

**Comment by catalintomai** [comment]
> @catalintomai as long as we are handling the validations (UI & Server) as part of the same release as this feature, I am fine with it. I have raised this issue separately here:
> https://mattermost.atlassian.net/browse/MM-22376

@srkgupta Thanks, I assigned that ticket to myself and address it as part of the same PR, now that we are in clear of the scope. Haven't verified, but I assume GoogleApps/Gitlab settings might have the same validation issue, so you might open a ticket for them as well (unrelated to current work).

**Comment by catalintomai** [comment]
> Thanks @catalintomai.
> 
> Tested the latest changes on localhost and the feature is working fine as expected. Just found one issue which occurred prior to this PR and have raised this issue separately.
> https://mattermost.atlassian.net/browse/MM-22405
> `Server should throw more detailed errors in the console & logs when it fails to start due to invalid config`
> 
> Existing testcases have been updated and new testcases have been added in the release spreadsheet. Approving the PR.

Hi @srkgupta - I updated the error strings to address https://mattermost.atlassian.net/browse/MM-22405 - can you validate? Thanks.

**Comment by jasonblais** [comment]
@catalintomai @srkgupta Do you know if any changes made here could break telemetry on whether office365 SSO is enabled? https://github.com/mattermost/mattermost-server/blob/master/app/diagnostics.go#L477

(My guess is not, but confirming)

**Comment by catalintomai** [comment]
> @catalintomai @srkgupta Do you know if any changes made here could break telemetry on whether office365 SSO is enabled? https://github.com/mattermost/mattermost-server/blob/master/app/diagnostics.go#L477
> 
> (My guess is not, but confirming)

@jasonblais, - I looked at the code and it is likely you are right - it should prob. be Office365Settings.SSOSettings().Enable. Please open a bug, we'll further investigate and address it this cycle. FYI: @srkgupta

**Pull request merged**